python/paddle/distributed/fleet/fleet.py

```diff
@@ -420,42 +426,168 @@ def broadcast_test(iteration, x, group):
+            for _ in range(iteration):
+                paddle.distributed.stream.reduce_scatter(
+                    output,
+                    x,
+                    op=paddle.distributed.ReduceOp.SUM,
+                    group=group,
+                    sync_op=True,
+                )
+            paddle.device.cuda.synchronize()
+            end_t = time.time()
+            return (end_t - start_t) / iteration
+
+        if hcg is None:
+            hcg = self.get_hybrid_communicate_group()
+
         dp_group = hcg.get_data_parallel_group()
         sharding_group = hcg.get_sharding_parallel_group()
+        mp_group = hcg.get_model_parallel_group()
         test_group = None
         if dp_group.nranks > 1:
             test_group = dp_group
         elif sharding_group.nranks > 1:
             test_group = sharding_group
-        else:
-            logger.warning(
-                f"hcg created with dp_degree: {dp_group.nranks} and sharding_degree: {sharding_group.nranks}, skipping perf test..."
-            )
-            return
+
         # test 1M ~ 1G
         nbytes = 1 << 20  # 1048576(1MB)
         final_nbytes = 1 << 30  # 1073741824(1GB)
         dtype = paddle.float32
+
+        # run once when test specific package size.
+        test_specific_size = False
+        for k, st in context.items():
+            if st[0] > 0:
+                test_specific_size = True
+                break
+
+        if test_specific_size:
+            test_comm = list(context.keys())
+
+        if len(test_comm) == 0:
+            return
+
         while nbytes <= final_nbytes:
             x = paddle.zeros([nbytes // 4], dtype=dtype)
             # warmup
             allreduce_test(iteration=10, x=x, group=test_group)
-            # test-allreduce
-            ret = allreduce_test(iteration=round, x=x, group=test_group)
-            logger.info(
-                f"[AllReduceTest] nbytes {nbytes}B test result: {ret} s/iter"
+
+            allreduce_size, allreduce_thres_time = context.get(
+                "allreduce", [nbytes, -1]
+            )
+            reduce_size, reduce_thres_time = context.get("reduce", [nbytes, -1])
+            broadcast_size, broadcast_thres_time = context.get(
+                "broadcast", [nbytes, -1]
             )
-            ret = reduce_test(iteration=round, x=x, group=test_group)
-            logger.info(
-                f"[ReduceTest] nbytes {nbytes}B test result: {ret} s/iter"
+            allgather_size, allgather_thres_time = context.get(
+                "allgather", [nbytes, -1]
             )
-            ret = broadcast_test(iteration=round, x=x, group=test_group)
-            logger.info(
-                f"[BroadcastTest] nbytes {nbytes}B test result: {ret} s/iter"
+            reduce_scatter_size, reduce_scatter_thres_time = context.get(
+                "reduce_scatter", [nbytes, -1]
             )
+
+            # inter machines
+            if "allreduce" in test_comm:
```

Comment:
Split this large function into small ones?

Reply:
DONE